ShowStatementTest.test_showJobsEmpty

PR builder against master (commit 58bcc8925f55ca806a180a003cf725ccbbbc9758)

https://jenkins.hazelcast.com/job/Hazelcast-pr-builder/9276/testReport/junit/com.hazelcast.jet.sql.impl/ShowStatementTest/test_showJobsEmpty/

<details><summary>Stacktrace:</summary>

```
java.lang.AssertionError: 

Expecting:
  <[Row{[testJob]}]>
to contain exactly (and in same order):
  <[]>
but some elements were not expected:
  <[Row{[testJob]}]>

	at com.hazelcast.jet.sql.SqlTestSupport.assertRowsOrdered(SqlTestSupport.java:279)
	at com.hazelcast.jet.sql.impl.ShowStatementTest.test_showJobsEmpty(ShowStatementTest.java:115)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

</details>

<details><summary>Standard output:</summary>

```
Finished Running Test: test_showJobs in 0.053 seconds.
Started Running Test: test_showJobsEmpty
BuildInfo right after test_showJobsEmpty(com.hazelcast.jet.sql.impl.ShowStatementTest): BuildInfo{version='5.1-SNAPSHOT', build='20220119', buildNumber=20220119, revision=3a2aafc, enterprise=false, serializationVersion=1}
11:06:00,577 ERROR || - [EventServiceImpl] hz.happy_bhabha.event-34 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] hz.happy_bhabha.event-34 caught an exception while processing:com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher@50270f54
java.lang.UnsupportedOperationException: SQL catalog must never be cleared - MapEvent{entryEventType=CLEAR_ALL, member=Member [127.0.0.1]:5701 - 78f8f138-1e6f-48e6-b7b7-d53d90d6da25 this, name='__sql.catalog', numberOfEntriesAffected=0}
	at com.hazelcast.jet.sql.impl.schema.TablesStorage$EntryListenerAdapter.mapCleared(TablesStorage.java:202) ~[classes/:?]
	at com.hazelcast.replicatedmap.impl.ReplicatedMapEventPublishingService.dispatchEvent(ReplicatedMapEventPublishingService.java:132) ~[hazelcast-5.1-SNAPSHOT.jar:5.1-SNAPSHOT]
	at com.hazelcast.replicatedmap.impl.ReplicatedMapService.dispatchEvent(ReplicatedMapService.java:232) ~[hazelcast-5.1-SNAPSHOT.jar:5.1-SNAPSHOT]
	at com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher.run(LocalEventDispatcher.java:64) ~[hazelcast-5.1-SNAPSHOT.jar:5.1-SNAPSHOT]
	at com.hazelcast.internal.util.executor.StripedExecutor$Worker.process(StripedExecutor.java:245) [hazelcast-5.1-SNAPSHOT.jar:5.1-SNAPSHOT]
	at com.hazelcast.internal.util.executor.StripedExecutor$Worker.run(StripedExecutor.java:228) [hazelcast-5.1-SNAPSHOT.jar:5.1-SNAPSHOT]
Hiccups measured while running test 'test_showJobsEmpty(com.hazelcast.jet.sql.impl.ShowStatementTest):'
11:06:00, accumulated pauses: 343 ms, max pause: 263 ms, pauses over 1000 ms: 0
```

</details>

--- a/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java
+++ b/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java
@@ -46,7 +46,6 @@ import com.hazelcast.jet.sql.impl.SqlPlanImpl.IMapSinkPlan;
 import com.hazelcast.jet.sql.impl.SqlPlanImpl.IMapUpdatePlan;
 import com.hazelcast.jet.sql.impl.SqlPlanImpl.SelectPlan;
 import com.hazelcast.jet.sql.impl.SqlPlanImpl.ShowStatementPlan;
-import com.hazelcast.jet.sql.impl.parse.SqlShowStatement.ShowStatementTarget;
 import com.hazelcast.jet.sql.impl.schema.TableResolverImpl;
 import com.hazelcast.map.IMap;
 import com.hazelcast.map.impl.EntryRemovingProcessor;
@@ -80,7 +79,6 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
-import java.util.Objects;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
@@ -290,15 +288,12 @@ public class PlanExecutor {
                 rows = catalog.getViewNames().stream();
                 break;
             case JOBS:
-                assert plan.getShowTarget() == ShowStatementTarget.JOBS;
                 NodeEngine nodeEngine = getNodeEngine(hazelcastInstance);
                 JetServiceBackend jetServiceBackend = nodeEngine.getService(JetServiceBackend.SERVICE_NAME);
-                rows = jetServiceBackend.getJobRepository().getJobRecords().stream()
-                        .map(record -> record.getConfig().getName())
-                        .filter(Objects::nonNull);
+                rows = jetServiceBackend.getJobRepository().getActiveJobNames().stream();
                 break;
             default:
-                throw new AssertionError("Unsupported SHOW statement target.");
+                throw new AssertionError("Unsupported SHOW statement target");
         }
         SqlRowMetadata metadata = new SqlRowMetadata(singletonList(new SqlColumnMetadata("name", VARCHAR, false)));
         InternalSerializationService serializationService = Util.getSerializationService(hazelcastInstance);

--- a/hazelcast/src/main/java/com/hazelcast/jet/impl/JobRepository.java
+++ b/hazelcast/src/main/java/com/hazelcast/jet/impl/JobRepository.java
@@ -90,6 +90,7 @@ import static java.util.Comparator.comparing;
 import static java.util.Objects.requireNonNull;
 import static java.util.concurrent.TimeUnit.HOURS;
 import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.stream.Collectors.toMap;
 import static java.util.stream.Collectors.toSet;
 
 public class JobRepository {
@@ -535,6 +536,18 @@ public class JobRepository {
         ids.addAll(jobRecordsMap().keySet());
         ids.addAll(jobResultsMap().keySet());
         return ids;
+    }
+
+    public Collection<String> getActiveJobNames() {
+        Map<Long, String> res = getJobRecords().stream()
+                .filter(record -> record.getConfig().getName() != null)
+                .collect(toMap(JobRecord::getJobId, record -> record.getConfig().getName()));
+        // When finalizing a job, we first create the JobResult, then delete the JobRecord.
+        // So it can happen that we saw a JobRecord for a completed job. Here we remove those.
+        for (JobResult result : getJobResults()) {
+            res.remove(result.getJobId());
+        }
+        return res.values();
     }
 
     public Collection<JobRecord> getJobRecords() {

